scorers.py

```diff
@@ -31,6 +36,81 @@ def score_sentence(self, ref, out):
     raise NotImplementedError("Sentence-level calculation is not implemented in BleuScorer as it is usually 0."
                               "Consider using SentenceBleuScorer (string sentbleu) instead.")
 
+  def cache_stats(self, cache_id, ref, out, weights=(0.25, 0.25, 0.25, 0.25)):
+    """
+    Cache sufficient statistics for caculating BLEU score
+
+    Args:
+      cache_id: Specifying which cache to use
+      ref: A reference corpus
+      out: An output corpus
+      weights: Weights for caculating BLEU score
+    """
+    def precision(ref, out, n):
+      out_ngram = ngram_utils.sent_ngrams_list(out, n)
+      ref_ngram = ngram_utils.sent_ngrams_list(ref, n)
+      out_cnt = Counter(out_ngram)
+      ref_cnt = Counter(ref_ngram)
+
+      num = 0
+      denom = 0
+      for ngram, o_cnt in out_cnt.items():
+        num += min(o_cnt, ref_cnt[ngram])
+        denom += o_cnt
+      denom = max(1, denom)
+
+      return num, denom
+  
+    self.cached_ref_len[cache_id] = defaultdict(lambda: 0)
```

Comment:
I don't like making the scorer stateful, as this seems like it could be a recipe for bugs. Instead, could you return a list of cached stats?